### ISSUE TITLE
Audio cracking/popping sound when pause or when dragging the progress bar

### ISSUE THREAD

**Issue by CCBIueX** [iina/iina#3617]
**System and IINA version:**

- macOS  12.1
- IINA  1.2.0 Build 129

**Expected behavior:** It is expected to not have a popping or cracking sound generated.

**Actual behavior:** A popping sound generated when pausing or dragging the progress bar

**Steps to reproduce:** **There is no issue when playing the video/audio on Quicktime Player or on youtube.com**, but the issue has **also been found in the mpv player** _(0.34.1 macOS builds by stolendata)_, after investigating I found the reason:
This is the wave form when you pause a video on IINA:
<img width="919" alt="截屏2022-01-19 下午7 36 37" src="https://user-images.githubusercontent.com/42627418/150123087-d65e6b75-e72a-4c3b-b4e8-69ca3c90ecae.png">

This is the wave form when you pause a video on Quicktime Player:
<img width="919" alt="截屏2022-01-19 下午7 41 23" src="https://user-images.githubusercontent.com/42627418/150123186-00aa13a2-1fc5-4b34-8dbf-a7c4f09f669c.png">

These two are not the same piece of audio but it describes the issue, Quicktime Player seems to give a really quick fade out effect when you pause that users doesn't notice (because it's too short) so the wave won't suddenly stop which will make the speaker generate a pop sound, and when you drag the progress bar, I guess Quicktime will also give a fade out and fade in effect (I am too lazy to do the experiment :P).

p.s. the results are also noticeable when you look at the spectrum diagram:

![屏幕录制2022-01-19 下午7 32 28](https://user-images.githubusercontent.com/42627418/150124149-bfb2c53b-ab35-40cd-b979-f4cb72fc4d97.png)

![屏幕录制2022-01-19 下午7 37 21](https://user-images.githubusercontent.com/42627418/150124161-07fddfaa-939b-4e9f-b454-20e23bb6b1b7.png)

- [ ] MPV does not have this problem.

**How often does this happen?**

**Comment by low-batt** [comment]
Really nice clear detailed report.

I see you have reproduced this with mpv and entered mpv issue https://github.com/mpv-player/mpv/issues/9730

As IINA uses mpv for audio IINA will need to wait for a fix from mpv.

The posted link to the mpv option is messed up. This is the referenced option in the [mpv manual](https://mpv.io/manual/master/#options-audio-stream-silence):

> **--audio-stream-silence=<yes|no>**
>
>    Cash-grab consumer audio hardware (such as A/V receivers) often ignore initial audio sent over HDMI. This can happen every time audio over HDMI is stopped and resumed. In order to compensate for this, you can enable this option to not to stop and restart audio on seeks, and fill the gaps with silence. Likewise, when pausing playback, audio is not stopped, and silence is played while paused. Note that if no audio track is selected, the audio device will still be closed immediately.
>
>    Not all AOs support this.
>
>>    Warning
>>
> >   This modifies certain subtle player behavior, like A/V-sync and underrun handling. Enabling this option is strongly discouraged.

Given that warning I feel this must be viewed as a workaround rather than a true solution.

This option can be enabled in IINA under Preferences / Advanced / Additional mpv options.

**Comment by ValZapod** [comment]
The only problem is that it is not bitperfect then. Do you think there is any way to do it by pausing on places where it will not cause in such pop?

**Comment by low-batt** [comment]
I consider this effect of the workaround to be a problem as well:
> Likewise, when pausing playback, audio is not stopped, and silence is played while paused.

This means that if I pause playing to do something else IINA will not go idle and will still be consuming energy in order to play silence, a problem for battery powered Macs. Undesirable and unexpected behavior.

Issue #3537 reports a problem with IINA consuming CPU while paused. That problem has been fixed in the develop branch. Using a build from that branch shows IINA idle when the video is paused:
```bash
low-batt@gag iina (develop $=)$ while true; do ps -c -o 'comm,%cpu,cputime' -u low-batt | grep IINA | grep -v grep; sleep 5; done
IINA               0.4   0:01.57
IINA               0.0   0:01.57
IINA               0.0   0:01.58
IINA               0.0   0:01.59
IINA               0.0   0:01.59
IINA               0.0   0:01.59
```

As expected from the documented behavior of `audio-stream-silence` IINA is no longer idle and is consuming CPU when the video is paused and that mpv property has been set to `yes`:
```bash
low-batt@gag iina (develop $=)$ while true; do ps -c -o 'comm,%cpu,cputime' -u low-batt | grep IINA | grep -v grep; sleep 5; done
IINA               0.1   0:02.10
IINA               0.1   0:02.12
IINA               0.2   0:02.14
IINA               0.2   0:02.15
IINA               0.2   0:02.16
IINA               0.2   0:02.17
```

I'm hoping for a solution from mpv along the lines of the fade in/out effect mentioned above and in mpv issue https://github.com/mpv-player/mpv/issues/9730

**Comment by JonnieCache** [comment]
Came here to express my support for this to be fixed.

Not totally sure of the relevance of the MPV issue tbh. This problem is specifically caused by the lack of a fade to or from silence.

These pops and clicks are caused by the loudspeaker driver (the "cone" part that gets moved back and forth by a magnet to create the soundwave) being asked to "snap" between zero and some other position instantaneously, something that it cannot physically do. It tries its best, and its failure is heard as a click. This is why the volume and frequency spectrum of the click is different each time  - it depends on the position and momentum of the driver when the signal goes to zero.

The minimum fade(out/in) length is thus determined by the samplerate, however for consumer playback (as opposed to musical) applications, a fade of ~20ms is long enough to hide any problems and will not be percieved by the user.

Maybe I've misunderstood the relevance of the upstream issue here, but hopefully this explanation can be useful.

It may be that in some circumstances these fades are being added somewhere else further up the audio pipeline, and other times they're not, as seen by the user in the MPV issue who found that enabling hardware decoding helped. IMO it would be best if IINA did its own fades in software, but obviously that may not be simple to implement :)

**Comment by low-batt** [comment]
Thank you very much for the detailed explanation of speaker behavior. Very helpful.

For audio/video support IINA uses a library from the [mpv](https://github.com/mpv-player/mpv) project. It is the code in that library that is decoding audio packets from the media being played and sending it to the audio devices. When IINA is commanded to pause playback it merely passes that instruction on to `mpv`. It is up to the `mpv` code to properly handle the audio devices.

This is why the mpv issue https://github.com/mpv-player/mpv/issues/9730 is important. The `mpv` project needs to address that issue and then release a new version of `mpv`. Once that is done then the IINA project will need to upgrade to the `mpv` version with the fix.

**Comment by krackers** [comment]
Nice graphs! I think there's actually two separate issues at play here:

* When using bluetooth headphones, as soon as the audio unit is stopped, macOS aggressively terminates the underlying bluetooth stream. Using audio unit reset does not trigger this behavior, since the system keeps the underlying stream active.
   -   Some types of audio devices produces artifacts simply when the stream is torn down (maybe powering off internal DAC), even when there is currently no playing output. 
   - You can replicate this type of behavior in quicktime by first pausing the stream then within a second quitting quicktime. I can reliably reproduce a pop, despite audio level already being 0 due to pause.
   -  I think a few years back the behavior in IINA (and mpv) was changed to match the strategy QuickTime uses, where AudioUnit is kept alive (via audio unit reset) for a few dozen seconds before ultimately being stopped (via au unit stop). 
   - So this should no longer be an issue for short pauses or seeking. For switching between tracks, this still always causes complete ao teardown. This can be avoided by using the gapless playback option.


* Separately, mpv (and IINA) do not "gracefully" fade in/out audio. As your graphs show, it is a hard stop, which can cause popping simply due to instantaneous change in amplitude. 
    - Different headphones might have different behavior here. I was not able to replicate this type of popping with my bluetooth headphones, perhaps they do some sort of filtering.
    - This behavior cannot be simply grafted onto the AO layer, because by the time the AO layer receives the "pause"/stop signal, mpv might have already stopped filling audio buffers so we may not have enough samples to fade gently. I
        * If you don't need a strict fade and just want to avoid pop, probably there is some digital-signal-processing approach to find a continuation of a signal that minimizes high frequency content. Just a linear ramp from the last value down to 0 I feel would still be noticeably jarring. I feel like you want a sinc-ish function, but i'm not sure how keeping only the positive half would affect things, there might be a better function to use here?
        * If you are lucky you might have a single full buffer worth of audio available to operate on though. I think this may allow you to fade for a short duration (5-10ms?)
           - I have not tested, but it could be that coreaudio automatically does this on AudioUnitReset rather than simply 0-flushing the buffer? That could be an alternate explanation why I didn't notice any issues.
    - mpv has an internal softvol mixer which can be used to replicate fade behavior via a lua script, to gently ramp the volume up/down on pause or track begin/end.

**Comment by low-batt** [comment]
On this:
> - I think a few years back the behavior in IINA (and mpv) was changed to match the strategy QuickTime uses, where AudioUnit is kept alive (via audio unit reset) for a few dozen seconds before ultimately being stopped (via au unit stop).

When it comes to audio behavior it is probably better to refer to mpv (and IINA), as the only role IINA plays is in setting mpv [audio](https://mpv.io/manual/stable/#audio) options to control the behavior of `libmpv`. It is `libmpv` that is decoding the audio stream and interfacing with macOS to drive the speakers.

The changes in mpv with respects to how audio was handled started with commit [39f7f83](https://github.com/mpv-player/mpv/commit/39f7f83351524cdbd87c5ae420b838530314b525). This commit changed mpv to use [AudioUnitReset](https://developer.apple.com/documentation/audiotoolbox/audiounitreset(_:_:_:)) instead of [AudioOutputUnitStop](https://developer.apple.com/documentation/audiotoolbox/audiooutputunitstop(_:)). This change was made to address mpv issue https://github.com/mpv-player/mpv/issues/9243, which had to do with lagging when using speakers connected over Bluetooth. The root cause was traced to [AudioOutputUnitStart](https://developer.apple.com/documentation/audiotoolbox/audiooutputunitstart(_:)) taking a noticiable about of time to start audio if the speakers were connected through Bluetooth.

_However_ that fix introduced the same problems that the workaround for this option of setting the mpv  [audio-stream-silence](https://mpv.io/manual/master/#options-audio-stream-silence) option to `yes` does, namely the macOS `coreaudiod` daemon consuming CPU and preventing sleep. This was reported in mpv issue https://github.com/mpv-player/mpv/issues/11617. The fix for that issue can be found in mpv PR https://github.com/mpv-player/mpv/pull/13663. This is the change that causes mpv to adopt the QuickTime strategy of calling [AudioOutputUnitStop](https://developer.apple.com/documentation/audiotoolbox/audiooutputunitstop(_:)) if audio remains idle for 7 seconds. 

On this:
> - mpv has an internal softvol mixer which can be used to replicate fade behavior via a lua script, to gently ramp the volume up/down on pause or track begin/end.

That would be nice. If mpv can be fixed with a lua script it should be possible to configure IINA to use that script as well.

I believe "softvol mixer" refers to what the mpv [volume](https://mpv.io/manual/stable/#options-volume) option calls "software volume":
> …
> This always controls the internal mixer (aka software volume).

The mpv issue https://github.com/mpv-player/mpv/issues/9730 has been closed as a duplicate of mpv issue https://github.com/mpv-player/mpv/issues/2515 which was entered in 2015 and is _still_ open. From the comments on that issue it seems some feel this should be the responsibility of the operating system.

I too find the cracking very irritating.

**Comment by krackers** [comment]
>refers to what the mpv [volume](https://mpv.io/manual/stable/#options-volume) option calls "software volume"

Yes, thank you for the archaeology of commits. You can see an example script of how to fade during seek in https://github.com/CounterPillow/mpv-quack/blob/master/quack.lua 

You could also do similar for next track/prev track/quit. It's a bit hacky solution but probably best that can be done at the moment. I'm not quite sure what was meant by OS handling it, since the OS doesn't have any more information than the AO layer about the samples. Perhaps he was referring to the solution of gracefully decaying down to 0 using some causal filter to remove as much high-freq content as possible.
